# scIsoSeq-Analysis-Tools

Recent technological advancements in high-throughput scRNA-Seq methods such Droplet-based single-cell sequencing allow thousands of cells to be captured and sequenced in a relatively short time frame and at a fraction of the cost. Such methods rely on capture of polyadenylated mRNA transcripts followed by cDNA synthesis, pooling, amplification, library construction, and Illumina 3’ cDNA sequencing. The combination of fragmentation and short-read sequencing fails to sufficiently sequence the V(D)J regions of rearranged TCR and BCR transcripts, which are located in the first 500 nucleotides at the 5’ end of the transcript. Consequently, 3’-tag scRNA-Seq platforms have limited application for determining clonotypic information from large numbers of lymphocytes. Variations on this approach employing 5’ cell barcodes enable the V(D)J sequences and global gene expression to be measured27, but don’t solve the need to integrate this information with the diversity of switching and alternative mRNA splicing involving the 3’ end of IGH mRNA. Recent advances in long-read sequencing technologies present a potential solution to the shortcomings of short-read sequencing. Full-length cDNA reads can encompass the entire sequence of BCR and TCR transcripts, but typically suffer from higher error rates and lower sequencing depth than short-read technologies.

High-throughput single-cell RNA sequencing is a powerful technique but only generates short reads from one end of a cDNA template, limiting the reconstruction of highly diverse sequences such as antigen receptors. 

https://github.com/shizhuoxing/BGI-Full-Length-RNA-Analysis-Pipeline/wiki
